The PFSignUpViewControler has only 4 fields, and it will be very help…

mmarshak opened the pull request:
…ful to add also a first name and last name as part of the registration.

I thought that this change maybe helpful for others as well.
so I added two more optional fields that appear before the username. People prefer to have their first name and last name before the other registration information.
Add the required to support also for PFSignUpView class.
All was tested .

nlutsenko commented:
Hey @mmarshak, thanks for the PR.

This is perfect and would be useful to everyone! 
I've highlighted few nits in code, as well as I have a concern here: the concept of firstName and lastName isn't quite localizable, since there are cultures that don't have either first name or last name.
I am open to ideas on how to improve it, but the first thing that comes to my mind is to use just a single field for both.

Bizzi-Body commented:
@nlutsenko - My recommendation is to use fields called "First Name" and "Family Name".  You then have enough information to present the correct name depending on local region/custom.

nlutsenko commented:
That's a good option. Though - if people say in India have only one name - it might confuse them. 
By a single field I meant something like 'Full Name'.

mmarshak commented:
Hi

In many signup forms for many apps there is a separation between the first name and last name, they are seperate field.

The two new fields are optional, so people do not have to use them.

We can make the text for the firstName field configured, so users in India can display instead just the text 'Name', and not show to the user the LastName field at all.

Regarding lastName, we can also allow the user to change the text from Last Name to Family name.

I can not access the code right now. I believe these two label an not readwrite in the header file. This mean that both these changes mean that the label need to be readwrite. 

Will this work ?

nlutsenko commented:
Can you please rebase the branch on the latest master and squash everything into a single commit?

nlutsenko commented:
What I meant was to rebase against master using something like:
`git rebase origin/master`
This would make sure that all the changes in these commits are put on top of the latest master.

Then for squashing - currently your pull request contains a lot of commits, where the actual change should be just 1 commit. To achieve the result you can use an interactive rebase.
The first result from Google http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html looks to be quite helpful.

Let me know if you have any questions or get blocked on something here.

mmarshak commented:
Ok, I did the rebase and the squashing. I hop I did it correctly.
Nikita, please let me know if you are all set to complete this pull request.

Thanks,
Marik

nlutsenko commented:
Hey @mmarshak. Almost there.
There are 3 commits in this PR, while it should be just 1 after all the squashing + rebasing.
Try setting the upstream of your local branch to ParseUI-iOS master, then doing an interactive rebase with `fixup` for all commits`, after it - simply push to your PR branch.

mmarshak commented:
Hi Nikita,

I tried doing it and get errors.

I will really appreciate you could could help and address this issue on your end.
Could you merge the code as is without asking me for those additional steps ?

Thanks,
Marik




mmarshak@Mariks-Mac-mini:~/Documents/ParseUI-iOS$ git rebase -i origin/master
error: could not apply d7ef5fe... The PFSignUpViewControler has only 4 fields, and it will be very helpful to add also a first name and last name as part of the registration.

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".

Could not apply d7ef5fe8cc4a2543e38abf490f3c117be4df661e... The PFSignUpViewControler has only 4 fields, and it will be very helpful to add also a first name and last name as part of the registration. I thought that this change maybe helpful for others as well. so I added two more optional fields that appear before the username. People prefer to have their first name and last name before the other registration information. Add the required to support also for PFSignUpView class. All was tested .

mmarshak@Mariks-Mac-mini:~/Documents/ParseUI-iOS$ git rebase -i origin/master

It seems that there is already a rebase-merge directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
	git rebase (--continue | --abort | --skip)
If that is not the case, please
	rm -fr "/Users/mmarshak/Documents/ParseUI-iOS/.git/rebase-merge"
and run me again.  I am stopping in case you still have something
valuable there.
mmarshak@Mariks-Mac-mini:~/Documents/ParseUI-iOS$ git rebase --continue
ParseUI/Classes/SignUpViewController/PFSignUpView.h: needs merge
ParseUI/Classes/SignUpViewController/PFSignUpView.m: needs merge
ParseUI/Classes/SignUpViewController/PFSignUpViewController.m: needs merge
You must edit all merge conflicts and then
mark them as resolved using git add



> On Jul 9, 2015, at 2:06 AM, Nikita Lutsenko <notifications@github.com> wrote:
> 
> Hey @mmarshak <https://github.com/mmarshak>. Almost there.
> There are 3 commits in this PR, while it should be just 1 after all the squashing + rebasing.
> Try setting the upstream of your local branch to ParseUI-iOS master, then doing an interactive rebase with fixup for all commits`, after it - simply push to your PR branch.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ParsePlatform/ParseUI-iOS/pull/97#issuecomment-119830275>.
>

mmarshak commented:
Hi Nikita,

Any suggestions ? Can we just merge the pull request as is ?

Thanks,
Marik

nlutsenko commented:
@mmarshak Let me do a fork from your fork and update it, so I can show you what I meant.
Sorry it takes a while, got tangled up in other things.

mmarshak commented:
Hi Nikita,

Thank you allot !

Marik


On Jul 16, 2015, at 6:37 PM, Nikita Lutsenko <notifications@github.com> wrote:

@mmarshak Let me do a fork from your fork and update it, so I can show you what I meant.
Sorry it takes a while, got tangled up in other things.

—
Reply to this email directly or view it on GitHub.

mmarshak commented:
Hi Nikita,

It seems that the pull request is still open.
I thought that in July you said that you will merge it.

Could you please merge it to the trunk ?

Thanks,
Marik

mmarshak commented:
Hi

Could one of the owner of the repository merge this pull request ?

Marik

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpView.m:
```diff
@@ -179,7 +235,7 @@ - (void)layoutSubviews {
         currentY = CGRectGetMaxY(frame);
     }
-
+    
```
Nit: Remove whitespace.

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpView.m:
```diff
@@ -187,7 +243,7 @@ - (void)layoutSubviews {
         currentY = CGRectGetMaxY(frame);
     }
-
+    
```
Nit: Remove whitespace.

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpViewController.m:
```diff
@@ -467,7 +503,7 @@ - (void)_updateSignUpViewContentOffsetAnimated:(BOOL)animated {
         // Scroll the view to keep fields visible
         CGFloat offsetForScrollingTextFieldToTop = CGRectGetMinY([self _currentFirstResponder].frame);
-
+        
```
Nit: Remove whitespace.

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpViewController.m:
```diff
@@ -261,7 +278,13 @@ - (void)_signUpAction {
         dictionary[PFSignUpViewControllerDelegateInfoAdditionalKey] = additional;
     }
-
+    if(firstName) {
```
Nit: Add space after `if` before `(` to be closer to overall coding style.

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpViewController.m:
```diff
@@ -261,7 +278,13 @@ - (void)_signUpAction {
+    if(firstName) {
+        dictionary[PFSignUpViewControllerDelegateInfoFirstNameKey] = firstName;
+    }
+    if(familyName) {
```
Nit: Add space after `if` before `(` to be closer to overall coding style.

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpView.m:
```diff
@@ -55,9 +55,49 @@ - (instancetype)initWithFields:(PFSignUpFields)otherFields {
+    }
+    
+    if (_fields & PFSignUpFieldsFamilyName) {
+        
```
Nit: Remove whitespace.

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpView.m:
```diff
@@ -155,7 +195,23 @@ - (void)layoutSubviews {
         currentY = floorf(CGRectGetMaxY(frame) + logoBottomInset);
     }
-
+    
```
Nit: Remove whitespace.

nlutsenko commented on ParseUI/Classes/SignUpViewController/PFSignUpView.m:
```diff
@@ -155,7 +195,23 @@ - (void)layoutSubviews {
+        
+        currentY = CGRectGetMaxY(frame);
+    }
+    
```
Nit: Remove spaces (line break is good though).

nlutsenko commented:
Another pass of changes and nits.
I am so sorry, but I still feel strong about having First Name + Family Name as 2 separate fields.
cc @grantland @richardjrossiii What do you guys think?

As I noted previously - please squash all the commits into a single one, so it's mergeable.
DIdn't get back to you earlier with the list how to get there, so here is step-by-step if you still need it ;)

To accomplish that in Terminal, inside ParseUI-iOS folder:
- Run `git rebase -i upstream/master` (replace upstream with the name you are using for ParseUI remote)
- You should see a list of commits appearing in your default editor
- Make every line start with `f` aka `fixup` instead of `pick`, except of the first commit in the list
- Close the editor saving the file
- You should see the rebase process happening and when it finishes - you can simply force push commits to this branch

mmarshak commented:
Hi Nikita,

On July 6 I fixed your previous comments, and they only thing you asked was to squash the commit.
At that time I tried to do it and I have an issue and then I asked you to do it. Now 2 months later you ask me again to do more changes, and the changes are to remove white spaces, man this is really nit picking. I am trying to submit a code to help the great good and you nit picking it. This does not really make people extremely motivated to contribute code ! Could you let it go this time ?

Regarding the first name and the last name debate, we already had it on Jun 17.
At that time Bizzi-Body suggested - "My recommendation is to use fields called "First Name" and "Family Name". You then have enough information to present the correct name depending on local region/custom." And I adopted this approach which you said it is a good one.
If people in India with to use 'Full Name' then with localizing the string (with NsLocalizesString) they can use instead of of the 'First name' the 'Full Name', and they can decide not to display the 'Last Name' field.
People in the US expect to see first name and last name. Looks at dozen of app in the apple store where user are required to register. There is always first name and last name, and both fields are separate. 
Having a single name called 'Full name' does not mean much in the US. People in the US will get confused what is mean ? Just first name, first name with middle name and last name or any other combination.

Regarding the ask for the squashing the code,  yesterday the CI said that the pull request is mergeable. So I do not get your point that the squash is required to make the code mergeable ?

Marik

nlutsenko commented:
The nits where still there before and from the history and seems like these ones never got fixed.
I am not picking on one specific thing, every single pull request goes through the same code review process, and we are aiming to keep the code clean and consistent. I am indeed nitpicking, that's why every single comment there has `nit` prefix. :grin: 

Rebasing - currently this pull request contains 13 commits, as well as some merge commits and things that don't quite go by themselves (e.g. there is a commit with the title `1`).

I sincerely hope this makes sense.
Is there any way I can help you get this code in?
I can merge it as is and clean up the nits myself, would that be an option?

The only thing that would help me probably is squashing the commits together into a single one.

mmarshak commented:
The local workspace that I used for the changes is messed up. So I did a new clone out of my repository.
When I run on it the command 
I get the following, which does not show any changes. So there is nothing to squash.
Could you take it from here and commit the changes to the trunk ?

Thanks,
Marik


Bellow is the output of - git rebase -i original /master

# Rebase 973cb27..973cb27 onto 973cb27 (       1 TODO item(s))
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out

nlutsenko commented:
I am sorry, but I simply can't make this an exception and merge these changes, since it's going to impact our commit history.
Looks like you are rebasing a wrong branch.
I would suggest trying to checkout the branch with these changes, then fetching from upstream (aka ParseUI-iOS/master) and then rebasing against that.

mmarshak commented:
Ok. I follow you exact suggestion of a new checkout, a fetch from the upstream. Which worked.
Then I went and did the following steps you recommend -

* Run git rebase -i upstream/master (replace upstream with the name you are using for ParseUI remote)
* You should see a list of commits appearing in your default editor
* Make every line start with f aka fixup instead of pick, except of the first commit in the list
* Close the editor saving the file

I changed all the pick with f except the first commit.
I saved the changes and closed the editor.
But I got the following error (it complain on the first commit in the file (the one I did not change to mixup). 

What I should do now ??? 

error: could not apply d7ef5fe... The PFSignUpViewControler has only 4 fields, and it will be very helpful to add also a first name and last name as part of the registration.

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".

Could not apply d7ef5fe8cc4a2543e38abf490f3c117be4df661e... The PFSignUpViewControler has only 4 fields, and it will be very helpful to add also a first name and last name as part of the registration. I thought that this change maybe helpful for others as well. so I added two more optional fields that appear before the username. People prefer to have their first name and last name before the other registration information. Add the required to support also for PFSignUpView class. All was tested .

nlutsenko commented:
This means that the first commit has a rebase conflict inside of it, which should be resolved, then added with `git add` and after it will allow you to `git rebase --continue`.

mmarshak commented:
I did the fetch and there were no conflict. So I am not sure where the conflict come from.
How you suggest me to find those conflicts ?

nlutsenko commented:
They are usually going to be marked in source files with `<<<` `>>>`, which will mark sections of code that are in your commit, as well as a piece of code that is conflicting in changes with it.

mmarshak commented:
I am still working on this.
I am getting an error from the rebalance about the strings file -
warning: Cannot merge binary files: ParseUI/Resources/Localization
/en.lproj/ParseUI.strings (HEAD vs. 646a57b... extend the demo)

Any suggestions ?

nlutsenko commented:
Hmm... Delete the old one, finish rebasing, build ParseUIDemo (which should trigger regeneration of the strings), amend to commit?

mmarshak commented:
I finished the rebasing without errors.
But when I do a push nothing happen, it does not update my repo in github. Any suggestion ?

nlutsenko commented:
Try force pushing aka `git push -f` instead of just `git push`

mmarshak commented:
The force push worked ! Thanks allot.

Now all the changes are in. Could you please merge the changes to the trunk ?

Thanks Marik

richardjrossiii commented:
There's still some code style changes that appear to have been missed in this diff, that will have to be fixed before we can merge it (we will never merge code that's against our style guidelines, sorry!). We do have an auto-format style, (`clang-format`) that you could use, if you wish.

I'm also not super keen on using 'First Name' and 'Last Name' separately, to me, it doesn't solve a concrete problem.

I think one question that has been failed to be asked so far is does this solve a use-case that's widespread enough to warrant being in the main ParseUI library?
  
 - Firstly, how many applications truly care about the user's legal given name, vs. a screen? Unless you're building a social media app, I would say that number is pretty small.
 - Secondly, there are so many different forms of name, and this hardly covers them all. For example: How would I register for this, when my legal name is 'Richard Ross III'? By keeping it as a simple single text field, we allow for more flexibility on name formats.

A good read on some of the issues with names can be found here: 
http://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/

I strongly recommend reading that, and for the reasons outlined in that article, I think it's wisest to keep it as a single field that we do no verification on (other than maybe checking for empty string), it allows the most compatibility for different cultures.

And personally, I would solve the 'full name vs family name' problem using simple localization strings, rather than re-architect the entire UI layer, which could potentially break existing users of ParseUI.

Also, this change doesn't exist on Android, so there is going to be additional work needed to be done on that from to keep feature parity.

/rant

mmarshak commented:
I disagree that there is a small number of apps that ask for separate first name and last name. I can list you many of those.
Also in many apps that require payment they care about separate names. 
Again if somebody does not want to use the First name and last name they do not have to use it, they can use localization to display instead of first name a full name.

Regarding the concern about the Android, I am an iOS developer not an Android developer. I think this is not reasonable to ask me to learn development for android if I want to add code to the repository for the benefit of many others.

Regarding some issue with style that you believe are there, Nikita said that he will take care of those. And what I need to do is to do the squash.

Thanks

mmarshak commented:
Guys any update with this pull request ?

Thanks,
Marik

hhanesand commented:
I don't think this is something that should be merged into the main branch, due to the complexity of names around the world. Much better to leave it as a single field.

@mmarshak If you really want to have a last name and first name, you could simply to the processing yourself before you upload the user to parse.

flovilmart commented:
We're taking over that repo and migrating over to the core SDK. PLease reopen a PR there if you still need the feature.
